### PR TITLE
SPARK-41973 Add HTML5 support

### PR DESCRIPTION
**Pull request by adriannowak**

### CODE REVIEW

File: appstore-metadata-service/src/main/java/com/lgi/appstore/metadata/api/converter/StringToApplicationTypeConverter.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+import static java.lang.String.format;
+import static java.util.Optional.ofNullable;
+
+@Component
+public class StringToApplicationTypeConverter implements Converter<String, ApplicationType> {
+
+    @Override
+    public ApplicationType convert(String name) {
+        try {
+            return ApplicationType.valueOf(ofNullable(name).orElse(""));
```
Comment (KrzysztofDzioba) [inline appstore-metadata-service/src/main/java/com/lgi/appstore/metadata/api/converter/StringToApplicationTypeConverter.java:34]: It's not clear what type of object is created here when static import is introduced. I like static imports, but for such factory method I would using avoid it

File: appstore-metadata-service/src/main/java/com/lgi/appstore/metadata/util/ApplicationTypeHelper.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+
+import static org.apache.commons.lang3.StringUtils.*;
+
+public final class ApplicationTypeHelper {
+    private ApplicationTypeHelper() {
+    }
+
+    public static ApplicationType fromValue(String type) {
+        if (isEmpty(type)) {
+            return null;
```
Comment (KrzysztofDzioba) [inline appstore-metadata-service/src/main/java/com/lgi/appstore/metadata/util/ApplicationTypeHelper.java:35]: ApplicationType is an enum, I don't think returning null is a good idea. I see null later on is mapped into native app. Maybe we can map it straightaway here?

File: appstore-metadata-service/src/main/java/com/lgi/appstore/metadata/util/ApplicationTypeHelper.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        }
+        try {
+            return ApplicationType.fromValue(type.toLowerCase(Locale.ROOT));
+        } catch (IllegalArgumentException e) {
```
Comment (KrzysztofDzioba) [inline appstore-metadata-service/src/main/java/com/lgi/appstore/metadata/util/ApplicationTypeHelper.java:39]: Let's add a log for this situation

File: appstore-metadata-service/src/main/java/com/lgi/appstore/metadata/util/ApplicationTypeHelper.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+            return null;
+        }
+    }
+
+    public static boolean isNativeApplication(ApplicationType applicationType, Collection<ApplicationType> webApplications) {
+        return applicationType == null || !webApplications.contains(applicationType);
```
Comment (KrzysztofDzioba) [inline appstore-metadata-service/src/main/java/com/lgi/appstore/metadata/util/ApplicationTypeHelper.java:45]: ApplicationType is an enum, I would expect it won't be null

File: appstore-metadata-service/src/main/java/com/lgi/appstore/metadata/util/ApplicationUrlService.java [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+import java.util.Objects;
+
+import static com.lgi.appstore.metadata.util.ApplicationTypeHelper.isNativeApplication;
+
+public class ApplicationUrlService {
+    private final ApplicationUrlCreator urlCreator;
+    private final Collection<ApplicationType> webApplications;
+
+    public ApplicationUrlService(ApplicationUrlCreator urlCreator, Collection<ApplicationType> webApplications) {
+        this.urlCreator = urlCreator;
```
Comment (KrzysztofDzioba) [inline appstore-metadata-service/src/main/java/com/lgi/appstore/metadata/util/ApplicationUrlService.java:34]:
```suggestion
        this.urlCreator = requireNonNull(urlCreator, "urlCreator");
```

File: appstore-metadata-service/src/main/java/com/lgi/appstore/metadata/util/ApplicationUrlService.java [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+        this.webApplications = webApplications;
```
Comment (KrzysztofDzioba) [inline appstore-metadata-service/src/main/java/com/lgi/appstore/metadata/util/ApplicationUrlService.java:35]:
```suggestion
        this.webApplications = requireNonNull(webApplications, "webApplications");
```

File: appstore-metadata-service/src/test/java/com/lgi/appstore/metadata/util/ApplicationUrlServiceTest.java [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+    private final ApplicationUrlService applicationUrlService = new ApplicationUrlService(urlCreator, WEB_APPLICATIONS);
+
+    @Test
+    void shouldGetUrlForWebApplication() {
+        // GIVEN
+        final var applicationUrlParams = createApplicationParams("application/vnd.rdk-app.html5");
+        doReturn("http://web").when(urlCreator).createApplicationUrl(Mockito.any(ApplicationUrlCreator.WebAppParams.class));
+
+        // WHEN
+        final var url = applicationUrlService.createApplicationUrlFromApplicationRecord(applicationUrlParams);
```
Comment (KrzysztofDzioba) [inline appstore-metadata-service/src/test/java/com/lgi/appstore/metadata/util/ApplicationUrlServiceTest.java:43]:
```suggestion
        final var actualUrl = applicationUrlService.createApplicationUrlFromApplicationRecord(applicationUrlParams);
```

File: appstore-metadata-service/src/test/java/com/lgi/appstore/metadata/util/ApplicationUrlServiceTest.java [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+    }
+
+    @Test
+    void shouldGetUrlForNativeApplication() {
+        // GIVEN
+        final var applicationUrlParams = createApplicationParams("any");
+        doReturn("http://native").when(urlCreator).createApplicationUrl(Mockito.any(ApplicationUrlCreator.NativeAppParams.class));
+
+        // WHEN
+        final var url = applicationUrlService.createApplicationUrlFromApplicationRecord(applicationUrlParams);
```
Comment (KrzysztofDzioba) [inline appstore-metadata-service/src/test/java/com/lgi/appstore/metadata/util/ApplicationUrlServiceTest.java:56]:
```suggestion
        final var actualUrl = applicationUrlService.createApplicationUrlFromApplicationRecord(applicationUrlParams);
```

File: appstore-metadata-service/src/main/java/com/lgi/appstore/metadata/util/ApplicationTypeHelper.java [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+            return null;
+        }
+        try {
+            return ApplicationType.fromValue(type.toLowerCase(Locale.ROOT));
+        } catch (IllegalArgumentException e) {
+            return null;
+        }
+    }
+
+    public static boolean isNativeApplication(ApplicationType applicationType, Collection<ApplicationType> webApplications) {
```
Comment (KrzysztofDzioba) [inline appstore-metadata-service/src/main/java/com/lgi/appstore/metadata/util/ApplicationTypeHelper.java:44]: isNativeOrOtherApplicationType

File: appstore-metadata-service/src/main/java/com/lgi/appstore/metadata/api/stb/PersistentAppsService.java [hunk 10]
```diff
@@ -68,10 +70,36 @@
...
+    @Override
+    public Optional<AppIdWithType> getApplicationType(String appId) {
+        return dslContext.select(APPLICATION.TYPE)
+                .from(APPLICATION)
+                .where(APPLICATION.ID_RDOMAIN.eq(appId))
+                .and(APPLICATION.VISIBLE.isTrue())
+                .orderBy(VERSION_SORT_FIELD)
+                .fetchStream()
+                .map(sqlRecord -> getAppIdWithType(appId, sqlRecord))
+                .findFirst();
```
Comment (den314) [inline appstore-metadata-service/src/main/java/com/lgi/appstore/metadata/api/stb/PersistentAppsService.java:89]: can we leverage limit 1 and not fetch all unecessary records from DB?

File: appstore-metadata-service/src/main/java/com/lgi/appstore/metadata/util/ApplicationTypeHelper.java [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+    private ApplicationTypeHelper() {
+    }
+
+    public static boolean isSupported(String applicationType) {
+        try {
+            fromValue(applicationType);
+            return true;
+        } catch (IllegalArgumentException illegalArgumentException) {
+            return false;
+        }
```
Comment (KrzysztofDzioba) [inline appstore-metadata-service/src/main/java/com/lgi/appstore/metadata/util/ApplicationTypeHelper.java:38]: We should avoid using exception-driven development

```suggestion
    return Arrays.stream(ApplicationType.values())
                    .map(ApplicationType::getValue)
                    .anyMatch(appType -> appType.equalsIgnoreCase(applicationType));
```

File: appstore-metadata-service/src/main/java/com/lgi/appstore/metadata/util/ApplicationTypeHelper.java [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+    }
+
+    public static boolean isNativeApplication(String applicationType, Collection<ApplicationType> webApplications) {
+        try {
+            final var type = fromValue(applicationType);
+            return !webApplications.contains(type);
+        } catch (IllegalArgumentException illegalArgumentException) {
+            return false;
+        }
```
Comment (KrzysztofDzioba) [inline appstore-metadata-service/src/main/java/com/lgi/appstore/metadata/util/ApplicationTypeHelper.java:47]:
```suggestion
return Arrays.stream(ApplicationType.values())
                    .map(ApplicationType::getValue)
                    .filter(appType -> appType.equals(applicationType))
                    .map(ApplicationType::fromValue)
                    .anyMatch(appType -> !webApplications.contains(appType));
```

File: appstore-metadata-service/src/test/java/com/lgi/appstore/metadata/api/service/BaseServiceTest.java [hunk 13]
```diff
@@ -75,14 +75,39 @@ protected ApplicationRecord createRandomApplicationRecord(MaintainerRecord maint
...
                                                               Feature feature) throws JsonProcessingException {
-        return createRandomApplicationRecord(maintainerRecord, localization, hardware, platform, dependency, feature, UUID.randomUUID().toString(),
-                String.valueOf(new Random().nextInt(200)), true);
+        return createRandomApplicationRecord(maintainerRecord,
+                localization,
+                hardware,
+                platform,
+                dependency,
+                feature,
+                UUID.randomUUID().toString(),
```
Comment (djasiok) [inline appstore-metadata-service/src/test/java/com/lgi/appstore/metadata/api/service/BaseServiceTest.java:84]: static import?

**Review by KrzysztofDzioba — APPROVED** [review]

File: appstore-metadata-service/src/main/java/com/lgi/appstore/metadata/api/stb/PersistentAppsService.java [hunk 14]
```diff
@@ -350,4 +375,19 @@ public Optional<StbApplicationDetails> getApplicationDetails(String appId, Strin
                                             .readValue(JsonObjectNames.PLATFORM, applicationMetadataRecord.get(APPLICATION.PLATFORM).data(), Platform.class)));
                 });
     }
+
+    private String createApplicationUrlFromApplicationRecord(Record applicationMetadataRecord, String platformName, String firmwareVer) {
```
[the remark this answers is not in the document]
Comment (adriannowak) [reply appstore-metadata-service/src/main/java/com/lgi/appstore/metadata/api/stb/PersistentAppsService.java:379]: I didn't want to mix these classes. I suggest keeping that duplication for now

File: appstore-metadata-service/src/main/java/com/lgi/appstore/metadata/api/stb/input/validator/PlatformAndVersionOptionalForWebValidator.java [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+import static org.apache.logging.log4j.util.Strings.isBlank;
+
+public class PlatformAndVersionOptionalForWebValidator {
+    private final Collection<ApplicationType> webApplications;
+
+    public PlatformAndVersionOptionalForWebValidator(Collection<ApplicationType> webApplications) {
+        this.webApplications = requireNonNull(webApplications, "webApplications");
+    }
+
+    public StbAppsListParams validate(StbAppsListParams params, AppIdWithType appIdWithType) {
```
[the remark this answers is not in the document]
Comment (adriannowak) [reply appstore-metadata-service/src/main/java/com/lgi/appstore/metadata/api/stb/input/validator/PlatformAndVersionOptionalForWebValidator.java:42]: I cannot agree more with you, refactored

**Review by den314 — APPROVED** [review]

**Review by pavlo-reshetniak — APPROVED** [review]

**Pull request merged**